### ISSUE TITLE
Unexpected "Cannot add supertype 'X' to sealed type 'Y'"

### ISSUE THREAD

**Issue by knutwannheden** [eclipse/xtext-core#515]
For the following parser rule I get an error marker "Unexpected "Cannot add supertype 'Statement' to sealed type 'FeatureCall'":

```xtext
CallOrAssignStatement returns Statement:
	FeatureCall
	(
		  {AssignStatement.target=current} ':=' value=INT
		| {CallStatement.expression=current}
	)
	';'
;
```

Since the rule should always return an instance of AssignStatement or CallStatement it seems wrong to me that Xtext requires FeatureCall to also be a subtype of the declared return type Statement. Or what am I missing here?

Full sample project attached: [mydsl.zip](https://github.com/eclipse/xtext-core/files/1400824/mydsl.zip)

**Comment by cdietrich** [comment]
hmmmm on which xtext version did you try that?

**Comment by cdietrich** [comment]
i assume xtext will never check this "always"

**Comment by cdietrich** [comment]
=> maybe really do it the classical way without ast rewrite

```
CallOrAssignStatement returns Statement:
	
	(
		  {AssignStatement} target=FeatureCall ':=' value=INT
		| {CallStatement} expression=FeatureCall
	)
	';'
;
```

**Comment by knutwannheden** [comment]
Yes, rewriting the grammar as you say will of course work. The validation still seems wrong to me.

**Comment by knutwannheden** [comment]
Intuitively I would expect the AST rewrite version to perform better because the parser can go right ahead and parse a FeatureCall without making any predictions and after that only one token lookahead is required to decide which alternative to take. But if that ends up making a difference in the larger scheme I wouldn't really know...

**Comment by cdietrich** [comment]
maybe @szarnekow can give us some insights

**Comment by szarnekow** [comment]
This is a broken validation in Xtext. We may want to enhance it to allow provable empty alternatives to have an impact on the type inference. Currently the logic was based on the assumption that error recovery may yield incomplete objects as the return value of a rule. If we have a current instance that is provable created / yielded by the last consumed text, we could relax / improve the validation.

**Comment by knutwannheden** [comment]
In my particular case (not reproducible with distilled example grammar) the proposed grammar rewrite doesn't work. It results in the "has non-LL(*) decision" error from ANTLR. This is what prompted the use of an AST rewrite rule in the first place.

As a workaround the Ecore model is now "polluted" with an unwanted supertype relationship.
